app/server/lib/Assistance.ts

```diff
@@ -2,116 +2,318 @@
  * Module with functions used for AI formula assistance.
  */
 
+import { AssistanceRequest, AssistanceResponse } from 'app/common/AssistancePrompts';
 import {delay} from 'app/common/delay';
+import { DocAction } from 'app/common/DocActions';
 import log from 'app/server/lib/log';
 import fetch from 'node-fetch';
 
 export const DEPS = { fetch };
 
-export async function sendForCompletion(prompt: string): Promise<string> {
-  let completion: string|null = null;
-  let retries: number = 0;
-  const openApiKey = process.env.OPENAI_API_KEY;
-  const model = process.env.COMPLETION_MODEL || "text-davinci-002";
+/**
+ * An assistant can help a user do things with their document,
+ * by interfacing with an external LLM endpoint.
+ */
+export interface Assistant {
+  apply(doc: AssistanceDoc, request: AssistanceRequest): Promise<AssistanceResponse>;
+}
 
-  while(retries++ < 3) {
-    try {
-      if (openApiKey) {
-        completion = await sendForCompletionOpenAI(prompt, openApiKey, model);
+/**
+ * Document-related methods for use in the implementation of assistants.
+ * Somewhat ad-hoc currently.
+ */
+export interface AssistanceDoc {
+  /**
+   * Generate a particular prompt coded in the data engine for some reason.
+   * It makes python code for some tables, and starts a function body with
+   * the given docstring.
+   * Marked "V1" to suggest that it is a particular prompt and it would
+   * be great to try variants.
+   */
+  assistanceSchemaPromptV1(options: {
+    tableId: string,
+    colId: string,
+    docString: string,
```

Comment:
It looks like the same type is also defined in-line in `ActiveDoc.ts`. Maybe we can extract it to a common interface.